analyses/response_analysis.py

```diff
@@ -0,0 +1,151 @@
+    plt.ylabel('upstream')
+    plt.tight_layout()
+    plt.savefig('upstream_dist_response_counts_500bins.pdf')
+    plt.close()
+
+    plt.figure(figsize=(10, 8))
+    ax = sns.distplot(upstream_response_counts_dask.compute(), bins=200, kde=False, color='blue',
+                      vertical=False, kde_kws={"clip": (0, 200)}, hist_kws={"range": (0, 200)})
+    ax.set_yscale('log')
+    plt.title('Upstream Distribution over responses, 200 bins')
+    plt.xlabel('responses')
+    plt.ylabel('upstream count')
+    plt.tight_layout()
+    plt.savefig('upstream_dist_response_counts_200bins.pdf')
+    plt.close()
+
+    patch_id_response_counts_dask = final[['patch_id', 'responses.resp_msg_id']].groupby('patch_id')[
+        'responses.resp_msg_id'].agg('count')
+
+    plt.figure(figsize=(12, 12))
+    ax = sns.distplot(patch_id_response_counts_dask.compute(), bins=500, kde=False, color='orange',
+                      vertical=False, kde_kws={"clip": (0, 500)}, hist_kws={"range": (0, 500)})
+    ax.set_yscale('log')
+    plt.xlabel('response count')
+    plt.ylabel('patches')
+    plt.title('Patch Distribution over number of responses')
+    plt.tight_layout()
+    plt.savefig('patch_dist_response_counts_500bins.pdf')
+    plt.close()
+
+    # Patch distribution by responses distinguished by with or without matching upstream
+    patch_responses_upstream_df = final[['patch_id', 'upstream', 'responses.resp_msg_id']]
+
+    patch_response_without_upstream_dask = patch_responses_upstream_df[
+        patch_responses_upstream_df['upstream'].isna() == True]
+    patch_response_without_upstream_grouped_dask = patch_response_without_upstream_dask.groupby('patch_id')[
+        'responses.resp_msg_id'].agg('count')
+
+    plt.figure(figsize=(11, 10))
+    ax = sns.distplot(patch_response_without_upstream_grouped_dask.compute(), kde=False, color='mediumvioletred', \
+                      vertical=False)
+    ax.set_yscale('log')
+    plt.title('Patch (without matching upstream) Distribution over number of responses')
+    plt.xlabel('response count')
+    plt.ylabel('patches')
+    plt.tight_layout()
+    plt.savefig('patch_without_upstream_dist_response_counts.pdf')
+    plt.close()
+
+    patch_response_with_upstream_dask = patch_responses_upstream_df[
+        patch_responses_upstream_df['upstream'].isna() == False]
+    patch_response_with_upstream_grouped_dask = patch_response_with_upstream_dask.groupby('patch_id')[
+        'responses.resp_msg_id'].agg('count')
+
+    plt.figure(figsize=(11, 10))
+    ax = sns.distplot(patch_response_with_upstream_grouped_dask.compute(), kde=False, color='olivedrab',
+                      vertical=False)
+    ax.set_yscale('log')
+    plt.title('Patch (with matching upstream) Distribution over number of responses')
+    plt.xlabel('response count')
+    plt.ylabel('patches')
+    plt.tight_layout()
+    plt.savefig('patch_with_upstream_dist_response_counts.pdf')
+    plt.close()
+
+    log.info("Saved plots, done!")
```

Comment:
Note: I didn't review the rest, I'd like to run it first.

Reply:
Ok